dialer.go

```diff
@@ -198,6 +203,29 @@ func (d *Dialer) LookupPartitions(ctx context.Context, network string, address s
 	return prt, err
 }
 
+// connectTLS returns a tls.Conn that has already completed the Handshake
+func (d *Dialer) connectTLS(ctx context.Context, conn net.Conn) (tlsConn *tls.Conn, err error) {
+	tlsConn = tls.Client(conn, d.TLS)
+	done := make(chan struct{})
+
+	go func() {
+		defer close(done)
+		err = tlsConn.Handshake()
```

Comment:
There's a race condition between `err = ctx.Err()` and `err = tlsConn.Handshake()` here.

One way to solve this would be to change the `done` channel to be of type `chan error` and publish the error returned by `tlsConn.Handshake()`. Then the select statement looks like this:

```go
select {
case <-ctx.Done():
    ...
case err = <-done:
}
```
That way there's only one goroutine writing to the `err` variable.